src/parquet/column/serialized-page.cc

```diff
@@ -45,23 +45,45 @@ SerializedPageReader::SerializedPageReader(std::unique_ptr<InputStream> stream,
   }
 }
 
-// TODO(wesm): this may differ from file to file
-static constexpr int DATA_PAGE_SIZE = 64 * 1024;
+static constexpr int DEFAULT_MAX_PAGE_HEADER_SIZE = 64 * 1024;
+static constexpr int MAX_PAGE_SIZE = 16 * 1024 * 1024;
 
 std::shared_ptr<Page> SerializedPageReader::NextPage() {
   // Loop here because there may be unhandled page types that we skip until
   // finding a page that we do know what to do with
   while (true) {
+    int bytes_available = 0;
     int bytes_read = 0;
-    const uint8_t* buffer = stream_->Peek(DATA_PAGE_SIZE, &bytes_read);
-    if (bytes_read == 0) {
-      return std::shared_ptr<Page>(nullptr);
-    }
-
-    // This gets used, then set by DeserializeThriftMsg
-    uint32_t header_size = bytes_read;
-    DeserializeThriftMsg(buffer, &header_size, &current_page_header_);
+    uint32_t header_size = 0;
+    const uint8_t* buffer;
+    int allowed_page_size = DEFAULT_MAX_PAGE_HEADER_SIZE;
+    std::stringstream ss;
+
+    // Page headers can be very large because of page statistics
+    // We try to deserialize a larger buffer progressively
+    // until a maximum allowed header limit
+    while (true) {
+      buffer = stream_->Peek(allowed_page_size, &bytes_available);
+      if (bytes_available == 0) {
+        return std::shared_ptr<Page>(nullptr);
+      }
 
+      // This gets used, then set by DeserializeThriftMsg
+      header_size = bytes_available;
+      try {
+        if (DeserializeThriftMsg(buffer, &header_size, &current_page_header_)) {
+          break;
+        }
+      } catch (std::exception& e) {
+        // Failed to deserialize. Double the allowed page header size and try again
+        ss << e.what();
+        allowed_page_size *= 2;
+        if (allowed_page_size > MAX_PAGE_SIZE) {
```

Comment:
@wesm I need your help in making this configurable. Can you please elaborate ? 
I patched your other comments. Let me know if you have more feedback. Thanks!